### PR TITLE
Dead code elimination

### PR DESCRIPTION
**Pull request by Protryon**
Belongs with module piercing as its base.

Implements basic intra-module dead code elimination.

### CODE REVIEW

File: src/main/java/com/shapesecurity/bandolier/es2017/transformations/DeadCodeElimination.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+		return Pair.of(removedImportSet, ImmutableList.of(item));
+	}
+
+	public static Pair<ImmutableSet<Pair<String, String>>, Module> removeAllUnusedDeclarations(@Nonnull Module module) {
+		DeadCodeElimination deadCodeElimination = new DeadCodeElimination(new ScopeLookup(ScopeAnalyzer.analyze(module)));
+		ImmutableList<Directive> directives = module.directives;
+		ImmutableList<ImportDeclarationExportDeclarationStatement> statements = module.items;
+		if (module.items.length == 1) { // IIFE
+			ImportDeclarationExportDeclarationStatement item = module.items.maybeHead().fromJust();
+			if (item instanceof CallExpression && ((CallExpression) item).arguments.length == 0 && ((CallExpression) item).callee instanceof FunctionExpression) {
```
Comment (michaelficarra) [inline src/main/java/com/shapesecurity/bandolier/es2017/transformations/DeadCodeElimination.java]: You have to also make sure there's no parameters being introduced.

File: src/main/java/com/shapesecurity/bandolier/es2017/transformations/DeadCodeElimination.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+				directives = module.directives.uniqByEquality().putAll(((FunctionExpression) ((CallExpression) item).callee).body.directives).toList();
```
Comment (michaelficarra) [inline src/main/java/com/shapesecurity/bandolier/es2017/transformations/DeadCodeElimination.java]: Save a reference to `((FunctionExpression) ((CallExpression) item).callee).body` to make this easier to read.

File: src/main/java/com/shapesecurity/bandolier/es2017/transformations/DeadCodeElimination.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+				statements = ((FunctionExpression) ((CallExpression) item).callee).body.statements.map(statement -> statement);
```
Comment (michaelficarra) [inline src/main/java/com/shapesecurity/bandolier/es2017/transformations/DeadCodeElimination.java]: Why this `map` of the identity?

File: src/main/java/com/shapesecurity/bandolier/es2017/transformations/DeadCodeElimination.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+			checkName(declaration.name.name);
+			Variable variable = lookup.findVariableDeclaredBy(declaration.name).fromJust();
+			if (variable.references.length == 0) {
+				return Pair.of(removedImportSet, ImmutableList.empty());
+			}
+		} else if (item instanceof VariableDeclarationStatement) {
+			//todo: array/object bindings?
+			ImmutableList<VariableDeclarator> declarators = ((VariableDeclarationStatement) item).declaration.declarators.filter(declarator -> {
+				if (declarator.init.isNothing() || declarator.init.fromJust() instanceof FunctionExpression|| declarator.init.fromJust() instanceof ClassExpression || declarator.init.fromJust() instanceof ArrowExpression) {
+					return true;
```
Comment (michaelficarra) [inline src/main/java/com/shapesecurity/bandolier/es2017/transformations/DeadCodeElimination.java]: Shouldn't this be `false`?

File: src/main/java/com/shapesecurity/bandolier/es2017/transformations/DeadCodeElimination.java (hunk 3 above)
Comment (Protryon) [reply src/main/java/com/shapesecurity/bandolier/es2017/transformations/DeadCodeElimination.java]: An implicit cast from `Statement` to `ImportDeclarationExportDeclarationStatement` takes place within the lambda, in effect casting `ImmutableList<Statement>` to `ImmutableList<ImportDeclarationExportDeclarationStatement>`.

Comment (michaelficarra) [reply src/main/java/com/shapesecurity/bandolier/es2017/transformations/DeadCodeElimination.java]: Unfortunately, that'll still actually do the walking/reconstruction. Explicitly cast the entire list.

Comment (Protryon) [reply src/main/java/com/shapesecurity/bandolier/es2017/transformations/DeadCodeElimination.java]: I resolved this by changing the types to begin with.

**Review by michaelficarra — APPROVED** [review]

**Pull request merged**